src/charm.py

```diff
@@ -177,116 +173,55 @@ def _on_get_certificate_action(self, event: ActionEvent) -> None:
             event: Juju event
         """
         hostname = event.params["hostname"]
-        tls_information = TLSInformation.from_charm(self)
-
-        tls_rel_data = tls_information.tls_requirer_integration.data[self.app]
-        if any(
-            not tls_rel_data.get(key)
-            for key in [f"certificate-{hostname}", f"ca-{hostname}", f"chain-{hostname}"]
-        ):
-            event.fail("Missing or incomplete certificate data")
-            return
-
-        event.set_results(
-            {
-                f"certificate-{hostname}": tls_rel_data[f"certificate-{hostname}"],
-                f"ca-{hostname}": tls_rel_data[f"ca-{hostname}"],
-                f"chain-{hostname}": tls_rel_data[f"chain-{hostname}"],
-            }
-        )
+        TLSInformation.from_charm(self, self.certificates)
```

Comment:
This line is not obvious, I assume there are relevant side-effects in it, but I think there should at least be a comment pointing this out (because no return value is used and the name of the function does not explain the side-effects). It would be better if we could have a solution without relying on side-effects, as it is easier to reason about.

Reply:
True, I missed that one

Reply:
we are replying on the `from_charm` method to check for the existence of the `certificates` integration and raises an exception which is handled by the `@validate_config_and_integration` decorator. I've added another class method called `validate` and used it here to be more explicit on what we're trying to do.